plasmapy/plasma/grids.py

```diff
@@ -1098,96 +1192,130 @@ def volume_averaged_interpolator(
+        #   near the boundary of the grid
+        bounding_cell_indices = np.empty((nparticles, 8, 3), dtype=np.int32)
+        lower_indices = np.where(
+            pos >= nearest_neighbor_pos,
+            nearest_neighbor_index,
+            nearest_neighbor_index - 1,
+        )
 
-                    d = np.abs(grid_pos - pos)
+        # populate x indices
+        bounding_cell_indices[:, 0:4, 0] = np.tile(
+            lower_indices[:, 0], (4, 1)
+        ).swapaxes(0, 1)
+        bounding_cell_indices[:, 4:, 0] = bounding_cell_indices[:, 0:4, 0] + 1
+
+        # populate y indices
+        bounding_cell_indices[:, [0, 1, 4, 5], 1] = np.tile(
+            lower_indices[:, 1], (4, 1)
+        ).swapaxes(0, 1)
+        bounding_cell_indices[:, [2, 3, 6, 7], 1] = (
+            bounding_cell_indices[:, [0, 1, 4, 5], 1] + 1
+        )
 
-                    # Fraction of cell volume that is closest to the
-                    # current point
-                    weight = (d[:, 0] * d[:, 1] * d[:, 2]) / cell_vol
-                    # weight = weight.to(u.dimensionless_unscaled)
-                    weight[out] = 0
-                    weight *= nan_mask
-                    weight = np.outer(weight, np.ones([nargs]))
+        # populate z indices
+        bounding_cell_indices[:, 0::2, 2] = np.tile(
+            lower_indices[:, 2], (4, 1)
+        ).swapaxes(0, 1)
+        bounding_cell_indices[:, 1::2, 2] = bounding_cell_indices[:, 0::2, 2] + 1
+
+        # Create a mask for cells whose locations would be off the grid,
+        # which occurs when the interpolation point is near the edge of the
+        # grid. These values will be weighted as zero later.
+        mask_cell_off = (
+            (bounding_cell_indices < 0).any(axis=2)
+            | (bounding_cell_indices[:, :, 0] >= n0)
+            | (bounding_cell_indices[:, :, 1] >= n1)
+            | (bounding_cell_indices[:, :, 2] >= n2)
+        )
 
-                    sum_value += weight * self._interp_quantities[x, y, z, :]
+        # Zero any out of bounds indices so IndexError is not raised
+        # during indexing.  This means an incorrect value will be retrieved
+        # but will not be used because of the zero weighting and the
+        # off the grid mask
+        bounding_cell_indices[mask_cell_off, :] = 0
+
+        # Calculate the volume of the overlap between the point volume
+        # and the volume of each of the surrounding vertices
+        lx = dx - np.abs(pos[:, None, 0] - ax0[bounding_cell_indices[..., 0]])
+        ly = dy - np.abs(pos[:, None, 1] - ax1[bounding_cell_indices[..., 1]])
+        lz = dz - np.abs(pos[:, None, 2] - ax2[bounding_cell_indices[..., 2]])
+        bounding_cell_weights = lx * ly * lz
+
+        # Set the weight for any off-grid vertices to zero
+        bounding_cell_weights[mask_cell_off] = 0.0
+        bounding_cell_weights[mask_particle_off, ...] = 0.0
+        bounding_cell_weights *= 1 / (dx * dy * dz)
```

Comment:
```suggestion
        norms = np.sum(bounding_cell_weights, axis=1)
        mask_norm_zero = norms == 0.0
        bounding_cell_weights[~mask_norm_zero] = (
            bounding_cell_weights[~mask_norm_zero, ...] / norms[~mask_norm_zero, None]
        )
```

I still feel this is the safer algorithm, but at this point they're essentially doing the same thing.   What actually might be good here is to issue a warning or exception if 

```python
not np.allclose(np.norm(bound_cell_weights[~mask_norm_zero]) / (dx * dy * dz), 1.0)
```

If this condition is not satisfied then the weight was not calculated correctly.